File: services/quant/app/market/instruments.py
"""Minimal instrument registry for the ingestion runners (QN-021).

Mirrors the Node BE-045 registry (`apis/node-api/src/market/instruments.ts`) for
the fields the Python side needs: the Twelve Data cross-check symbol and the pip
size. Energy CFDs have no reliable Twelve Data symbol, so their cross-check is
skipped (symbol = None).
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class InstrumentInfo:
    name: str
    twelve_data_symbol: str | None
    pip: float


_REGISTRY: dict[str, InstrumentInfo] = {
    "EUR_USD": InstrumentInfo("EUR_USD", "EUR/USD", 0.0001),
    "GBP_USD": InstrumentInfo("GBP_USD", "GBP/USD", 0.0001),
    "USD_JPY": InstrumentInfo("USD_JPY", "USD/JPY", 0.01),
    "USD_CHF": InstrumentInfo("USD_CHF", "USD/CHF", 0.0001),
    "AUD_USD": InstrumentInfo("AUD_USD", "AUD/USD", 0.0001),
    "USD_CAD": InstrumentInfo("USD_CAD", "USD/CAD", 0.0001),
    "NZD_USD": InstrumentInfo("NZD_USD", "NZD/USD", 0.0001),
    "XAU_USD": InstrumentInfo("XAU_USD", "XAU/USD", 0.01),
    "WTICO_USD": InstrumentInfo("WTICO_USD", None, 0.01),
    "BCO_USD": InstrumentInfo("BCO_USD", None, 0.01),
}


def get_instrument(name: str) -> InstrumentInfo:
    """Registry lookup; falls back to a 4-dp pip and no cross-check for unknowns."""
    return _REGISTRY.get(name, InstrumentInfo(name, None, 0.0001))

File: apis/node-api/src/market/instruments.ts
import type { Instrument, InstrumentInfo } from '@fx/types';

/**
 * BE-045 — static instrument registry (there is no `instruments` DB table in
 * Step 1.4). ~8 instruments in scope: FX majors + XAU/USD + WTI/Brent
 * (System Design scope guardrails). Broker symbol mappings here seed QN-033's
 * fuller symbol-mapping table in Phase 2. `pipLocation` is the power-of-ten of
 * one pip (EUR_USD pip 0.0001 → -4; JPY/metals/energy → -2).
 */
export const INSTRUMENTS: readonly InstrumentInfo[] = [
  fx('EUR_USD', 'EUR/USD', 'USD', -4),
  fx('GBP_USD', 'GBP/USD', 'USD', -4),
  fx('USD_JPY', 'USD/JPY', 'JPY', -2),
  fx('USD_CHF', 'USD/CHF', 'CHF', -4),
  fx('AUD_USD', 'AUD/USD', 'USD', -4),
  fx('USD_CAD', 'USD/CAD', 'CAD', -4),
  fx('NZD_USD', 'NZD/USD', 'USD', -4),
  {
    name: 'XAU_USD',
    displayName: 'Gold',
    kind: 'metal',
    oandaSymbol: 'XAU_USD',
    twelveDataSymbol: 'XAU/USD',
    pipLocation: -2,
    quoteCurrency: 'USD',
    tradeable: true,
  },
  {
    name: 'WTICO_USD',
    displayName: 'WTI Crude Oil',
    kind: 'energy',
    oandaSymbol: 'WTICO_USD',
    // Twelve Data free tier does not reliably carry OANDA's CFD oil symbol;
    // cross-check is skipped for energy until a vendor mapping is confirmed.
    twelveDataSymbol: null,
    pipLocation: -2,
    quoteCurrency: 'USD',
    tradeable: true,
  },
  {
    name: 'BCO_USD',
    displayName: 'Brent Crude Oil',
    kind: 'energy',
    oandaSymbol: 'BCO_USD',
    twelveDataSymbol: null,
    pipLocation: -2,
    quoteCurrency: 'USD',
    tradeable: true,
  },
];

function fx(
  name: string,
  twelveDataSymbol: string,
  quoteCurrency: string,
  pipLocation: number,
): InstrumentInfo {
  return {
    name,
    displayName: name.replace('_', '/'),
    kind: 'fx_major',
    oandaSymbol: name,
    twelveDataSymbol,
    pipLocation,
    quoteCurrency,
    tradeable: true,
  };
}

const BY_NAME: ReadonlyMap<string, InstrumentInfo> = new Map(INSTRUMENTS.map((i) => [i.name, i]));

/** Lookup by canonical name; undefined for unknown/unsupported instruments. */
export function getInstrument(name: string): InstrumentInfo | undefined {
  return BY_NAME.get(name);
}

/** True when `name` is a configured, in-scope instrument. */
export function isKnownInstrument(name: string): name is Instrument {
  return BY_NAME.has(name);
}

/** One pip expressed as an absolute price increment, e.g. EUR_USD → 0.0001. */
export function pipSize(info: InstrumentInfo): number {
  return 10 ** info.pipLocation;
}
